libraries/mediaviewer/impl/src/main/kotlin/io/element/android/libraries/mediaviewer/impl/gallery/MediaTimeline.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2025 New Vector Ltd.
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+ * Please see LICENSE files in the repository root for full details.
+ */
+
+package io.element.android.libraries.mediaviewer.impl.gallery
+
+import com.squareup.anvil.annotations.ContributesBinding
+import io.element.android.libraries.di.RoomScope
+import io.element.android.libraries.di.SingleIn
+import io.element.android.libraries.matrix.api.core.EventId
+import io.element.android.libraries.matrix.api.core.UniqueId
+import io.element.android.libraries.matrix.api.room.MatrixRoom
+import io.element.android.libraries.matrix.api.timeline.Timeline
+import kotlinx.collections.immutable.persistentListOf
+import javax.inject.Inject
+
+interface MediaTimeline {
+    suspend fun getTimeline(): Result<Timeline>
+    val cache: GroupedMediaItems?
+    fun orCache(data: GroupedMediaItems): GroupedMediaItems
+}
+
+/**
+ * A timeline holder that can be used by the gallery and the media viewer.
+ * When opening the Media Viewer, if the held timeline knows the Event, it will
+ * be used, else a FocusedMediaTimeline will be used.
+ */
+@SingleIn(RoomScope::class)
+@ContributesBinding(RoomScope::class)
+class LiveMediaTimeline @Inject constructor(
+    private val room: MatrixRoom,
+) : MediaTimeline {
+    private var timeline: Timeline? = null
```

Comment:
This is not thread safe, you should at least try to synchronise with a mutex

Reply:
Done in [c059c3f](https://github.com/element-hq/element-x-android/pull/4205/commits/c059c3fd2770eb6aaf52adb6813ee29e7111ded0)